src/nnet3/nnet-discriminative-training.h

```diff
@@ -0,0 +1,146 @@
+// nnet3/nnet-discriminative-training.h
+
+// Copyright 2012-2015   Johns Hopkins University (author: Daniel Povey)
+//           2014-2015   Vimal Manohar
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef KALDI_NNET3_NNET_DISCRIMINATIVE_TRAINING_H_
+#define KALDI_NNET3_NNET_DISCRIMINATIVE_TRAINING_H_
+
+#include "nnet3/nnet-example.h"
+#include "nnet3/nnet-computation.h"
+#include "nnet3/nnet-compute.h"
+#include "nnet3/nnet-optimize.h"
+#include "nnet3/nnet-discriminative-example.h"
+#include "nnet3/nnet-training.h"
+#include "nnet3/discriminative-training.h"
+
+namespace kaldi {
+namespace nnet3 {
+
+struct NnetDiscriminativeTrainingOptions {
+  NnetTrainerOptions nnet_config;
+  discriminative::DiscriminativeTrainingOptions discriminative_training_config;
+  discriminative::DiscriminativeTrainingStatsOptions discriminative_training_stats_config;
+
+  bool apply_deriv_weights;
+
+  NnetDiscriminativeTrainingOptions(): apply_deriv_weights(true) { }
+
+  void Register(OptionsItf *opts) {
+    nnet_config.Register(opts);
+    discriminative_training_config.Register(opts);
+    discriminative_training_stats_config.Register(opts);
+    opts->Register("apply-deriv-weights", &apply_deriv_weights,
+                   "If true, apply the per-frame derivative weights stored with "
+                   "the example (you'll normally want to leave this as true.");
+  }
+};
+
+// This struct is used in multiple nnet training classes for keeping
+// track of objective function values.
+// Also see struct AccuracyInfo, in nnet-diagnostics.h.
+struct DiscriminativeObjectiveFunctionInfo {
+  int32 current_phase;
+
+  double tot_aux_objf;
```

Comment:
I'm not sure if you changed this code since my last comments, or if this is different class?
What is tot_aux_objf needed for?  My suspicion is that either it is not needed, or you could give it a more specific name to indicate its exact function.
And I don't think we need a struct to wrap a struct here-- please just have one level of struct for this.